source/fe/mapping_manifold.cc

```diff
@@ -0,0 +1,1308 @@
+    {
+      const UpdateFlags update_flags = data.update_each;
+
+      AssertDimension(data.vertices.size(), GeometryInfo<dim>::vertices_per_cell);
+
+      if (update_flags & update_quadrature_points)
+        {
+          for (unsigned int point=0; point<quadrature_points.size(); ++point)
+            {
+              quadrature_points[point] = data.manifold->
+                                         get_new_point(Quadrature<spacedim>(data.vertices,
+                                                       data.cell_manifold_quadrature_weights[point+data_set]));
+            }
+        }
+    }
+
+
+    /**
+     * Update the co- and contravariant matrices as well as their determinant, for the cell
+     * described stored in the data object, but only if the update_flags of the @p data
+     * argument indicate so.
+     */
+    template <int dim, int spacedim>
+    void
+    maybe_update_Jacobians (const typename dealii::QProjector<dim>::DataSetDescriptor          data_set,
+                            const typename dealii::MappingManifold<dim,spacedim>::InternalData      &data)
+    {
+      const UpdateFlags update_flags = data.update_each;
+
+      if (update_flags & update_contravariant_transformation)
+        {
+          const unsigned int n_q_points = data.contravariant.size();
+
+          std::fill(data.contravariant.begin(), data.contravariant.end(),
+                    DerivativeForm<1,dim,spacedim>());
+
+          // Cache of weights used to compute points on the reference cell
+          std::vector<double> weights(GeometryInfo<dim>::vertices_per_cell);
+
+          AssertDimension(GeometryInfo<dim>::vertices_per_cell,
+                          data.vertices.size());
+          for (unsigned int point=0; point<n_q_points; ++point)
+            {
+              // Start by figuring out how to compute the direction in
+              // the reference space:
+              const Point<dim> &p = data.quad.point(point+data_set);
+
+              // And get its image on the manifold:
+              const Point<spacedim> P = data.manifold->
+                                        get_new_point(Quadrature<spacedim>(data.vertices,
+                                                      data.cell_manifold_quadrature_weights[point+data_set]));
+
+              // Always get the maximum length from the point to the
+              // boundary of the reference element, to compute the
+              // tangent vectors from the Manifold object
+              for (unsigned int i=0; i<dim; ++i)
+                {
+                  Point<dim> ei = Point<dim>::unit_vector(i);
+                  double ai = ei*p;
+                  Assert(ai >=0, ExcInternalError("Was expecting a quadrature point "
+                                                  "inside the unit reference element."));
+                  Point<dim> np(ai > .5 ? p-ai *ei : p+(1-ai)*ei);
```

Comment:
so if I see this right, you are constructing a point on the boundary of the reference cell where you choose the boundary (in direction `i`) that is further away. Is this correct? Either way, maybe provide a comment to the effect of what you're trying to do here.

The constructor argument is pretty long. I think it would be easier to read if you just said
```
  Point<dim> np = (...);
```
to make clear that it's just a copy assignment. I believe you can also make the object `const`.